<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-0.9.584">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="Dr.&nbsp;Gilbert and Dr.&nbsp;Duryea">

<title>Finding the Replication Origin, Part IV</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
span.underline{text-decoration: underline;}
div.column{display: inline-block; vertical-align: top; width: 50%;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>


<script src="ReplicationOrigin_PartIV_files/libs/clipboard/clipboard.min.js"></script>
<script src="ReplicationOrigin_PartIV_files/libs/quarto-html/quarto.js"></script>
<script src="ReplicationOrigin_PartIV_files/libs/quarto-html/popper.min.js"></script>
<script src="ReplicationOrigin_PartIV_files/libs/quarto-html/tippy.umd.min.js"></script>
<script src="ReplicationOrigin_PartIV_files/libs/quarto-html/anchor.min.js"></script>
<link href="ReplicationOrigin_PartIV_files/libs/quarto-html/tippy.css" rel="stylesheet">
<link href="ReplicationOrigin_PartIV_files/libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="ReplicationOrigin_PartIV_files/libs/bootstrap/bootstrap.min.js"></script>
<link href="ReplicationOrigin_PartIV_files/libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="ReplicationOrigin_PartIV_files/libs/bootstrap/bootstrap.min.css" rel="stylesheet">

  <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js" type="text/javascript"></script>

</head>

<body class="fullcontent">

<div id="quarto-content" class="page-columns page-rows-contents page-layout-article">

<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">Finding the Replication Origin, Part IV</h1>
</div>



<div class="quarto-title-meta">

    <div>
    <div class="quarto-title-meta-heading">Author</div>
    <div class="quarto-title-meta-contents">
             <p>Dr.&nbsp;Gilbert and Dr.&nbsp;Duryea </p>
          </div>
  </div>
    
    
  </div>
  

</header>

<section id="preliminary-thoughts" class="level2">
<h2 class="anchored" data-anchor-id="preliminary-thoughts">Preliminary Thoughts</h2>
<p>In all of our work so far, we’ve taken a fairly naive approach to discovering the replication origin. We haven’t leveraged what biology can tell us about our search. Let’s remember some biology and use it to our advantage.</p>
<section id="the-machinery-of-genome-replication" class="level3">
<h3 class="anchored" data-anchor-id="the-machinery-of-genome-replication">The Machinery of Genome Replication</h3>
<p>Recall from gen-bio that DNA replication is asymmetric, meaning there is a leading strand and a lagging strand. The leading strand is replicated as one continuous section because it is oriented in the correct direction for the replication machinery. However, the lagging strand is pointed in the “wrong” direction, so it is replicated more slowly in chunks that are called Okazaki fragments.</p>
<p>One thing we did not mention in gen-bio is that this actually exposes the lagging strand to more potential errors during replication. This is because the lagging strand is actually signal stranded for more time and is more vulnerable to mutation during this time. One type of mutation that can occur is called deamination. This is where cytosine (C) mutates into thymine (T). Take a minute to review the chemical structure of both of these nucleotides. Can you see how cytosine could turn into T?</p>
<p>After this first mutation occurs (C mutating into T), we would have T across from G at this position of the double helix. When this genome is replicated again, the DNA polymerase would correct this “mismatch” so that the G would be replaced by A, to follow the rules of complementary base pairing. Thus, the overall result of deamination is a decrease in the frequency of G in that section of the genome.</p>
<p>Now that we know about deamination, let’s see if we can use it to our advantage. It turns out that looking at the cumulative difference in observed frequencies of guanine and cytosine along the genome can be really helful in identifying the location where we should begin searching for the replication origin.</p>
<p><strong>Definition:</strong> Let’s define <code>SKEW(genome)</code> to be a function evaluated along a given genome, with a resulting value reported at each nucleotide. We select an arbitrary starting nucleotide (position 0), and report the <em>skew</em> at nucleotide <em>i</em> along the genome as the difference between the total number of observed guanine and total number of observed cytosine between position <em>0</em> and position <em>i</em>. We set <span class="math inline">\(SKEW_0\left(genome\right) = 0\)</span> and can compute <span class="math inline">\(SKEW_{i+1}\left(genome\right)\)</span> as follows:</p>
<p><span class="math display">\[SKEW_{i+1} = \left\{\begin{array}{l} SKEW_{i}\left(genome\right) + 1, \text{ if } genome_{i} = G\\
SKEW_{i}\left(genome\right) - 1, \text{ if } genome_{i} = C\\
SKEW_{i}\left(genome\right), \text{ otherwise} \end{array}\right.\]</span></p>
<hr>
<p><strong>Challenge 1:</strong> Build a function to compute <code>SKEW()</code>.<br>
&gt; <strong>Input:</strong> A <code>genome</code><br>
&gt; <strong>Output:</strong> A list, <code>skew</code> such that <code>skew[i]</code> contains the value of the <code>SKEW()</code> function at nucleotide position <em>i</em> in <code>genome</code>.</p>
<hr>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="co">#Complete the challenge here</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Does your <code>SKEW()</code> function work? How do you know? If you are confident, use <code>SKEW()</code> to compute the <em>skew</em> along the E. Coli genome. Store the results into a list called <code>e_coli_skew</code></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="co">#Compute e_coli_skew here</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Use the code below to plot thew skew along the E. Coli genome.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>() <span class="sc">+</span></span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_line</span>(<span class="at">mapping =</span> <span class="fu">aes</span>(<span class="at">x =</span><span class="dv">1</span><span class="sc">:</span><span class="fu">length</span>(e_coli_skew), <span class="at">y =</span> e_coli_skew))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Thats interesting and quite informative. The biological theory suggests that we should look for the replication origin near the minimum of the skew function.</p>
<hr>
<p><strong>Challenge 2:</strong> Create a new function <code>SKEW_min()</code> so that it returns the location(s) of the minimum <em>skew</em> along the genome rather than returning a full list of skew values.</p>
<hr>
<div class="cell">
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="co">#Complete the challenge here</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Use your new <code>SKEW_min()</code> function to solve the <a href="http://rosalind.info/problems/ba1f/">Minimum Skew Problem</a> on Rosalind.</p>
<hr>
<p><strong>Challenge 3:</strong> Solve the Minimum Skew Problem on Rosalind.</p>
<hr>
<div class="cell">
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="co">#Complete the challenge here</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
</section>
<section id="summary-and-debrief" class="level2">
<h2 class="anchored" data-anchor-id="summary-and-debrief">Summary and Debrief</h2>
<p>Similar to the last notebook, I’ll need to come back and write this section. We may need some more scaffolding here as well.</p>
</section>

</main>
<!-- /main column -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    target: function(trigger) {
      return trigger.previousElementSibling;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    setTimeout(function() {
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      let href = ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const cites = ref.parentNode.getAttribute('data-cites').split(' ');
    tippyHover(ref, function() {
      var popup = window.document.createElement('div');
      cites.forEach(function(cite) {
        var citeDiv = window.document.createElement('div');
        citeDiv.classList.add('hanging-indent');
        citeDiv.classList.add('csl-entry');
        var biblioDiv = window.document.getElementById('ref-' + cite);
        if (biblioDiv) {
          citeDiv.innerHTML = biblioDiv.innerHTML;
        }
        popup.appendChild(citeDiv);
      });
      return popup.innerHTML;
    });
  }
});
</script>
</div> <!-- /content -->



</body></html>